Documentation/botr/shared-generics.md

```diff
@@ -0,0 +1,185 @@
+
+# Dynamic Dictionary Expansion
+
+### History
+
+Before the dynamic dictionary expansion feature, dictionary layouts were organized into buckets (a linked list of fixed-size `DictionaryLayout` structures). The size of the initial layout bucket was always fixed to some number which was computed based on some heuristics for generic types, and always fixed to 4 slots for generic methods. The generic types and methods also had fixed-size generic dictionaries which could be used for lookups (also known as "fast lookup slots").
+
+When a bucket gets filled with entries, we would just allocate a new `DictionaryLayout` bucket, and add it to the list. The problem however is that we couldn't resize the generic dictionaries of types or methods, because they have already been allocated with a fixed size, and the JIT does not support generating instructions that could indirect into a linked-list of dictionaries. Given that limitation, we could only lookup a generic dictionary for a fixed number of values (the ones associated with the entries of the first `DictionaryLayout` bucket), and were forced to go through a slower runtime helper for additional lookups.
+
+This was acceptable, until we introduced the [ReadyToRun](https://github.com/dotnet/coreclr/blob/master/Documentation/botr/readytorun-overview.md) and the Tiered Compilation technologies. Slots were getting assigned quickly when used by ReadyToRun code, and when the runtime decided re-jitted certain methods for better performance, it could not in some cases find any remaining "fast lookup slots", and was forced to generate code that goes through the slower runtime helpers. This ended up hurting performance in some scenarios, and a decision was made to not use the fast lookup slots for ReadyToRun code, and instead keep them reserved for re-jitted code. This decision however hurt the ReadyToRun performance, but it was a necessary compromise since we cared more about re-jitted code throughput over R2R throughput.
+
+For this reason, the dynamic dictionary expansion feature was introduced.
+
+### Description and Algorithms
+
+The feature is simple in concept: change dictionary layouts from a linked list of buckets into dynamically expandable arrays instead. Sounds simple, but great care had to be taken when impementing it, because:
+- We can't just resize `DictionaryLayout` structures alone. If the size of the layout is larger than the size of the actual generic dictionary, this would cause the JIT to generate indirection instructions that do not match the size of the dictionary data, leading to access violations.
+- We can't just resize generic dictionaries on types and methods:
+    - For types, the generic dictionary is part of the `MethodTable` structure, which can't be reallocated (already in use by managed code)
```

Comment:
This does not sound accurate. Generic dictionary is not inline in the `MethodTable` structure. It is pointed to from the `MethodTable` structure.

I do not see much difference between the `MethodDesc` and `MethodTable` situation.

Reply:
It is part of the MethodTable, unlike MethodDescs: https://github.com/dotnet/coreclr/blob/master/src/vm/generics.cpp#L288

Reply:
This is just a micro-optimization to reduce number of calls into LoaderAllocator. The dictionary can be separate allocation and everything would still work just fine.

Similarly, we could apply the same micro-optimization to MethodDesc too if we wanted to.

Reply:
Ok I see what you mean. I'll reword this in the document.